kennedykori/SudokuLib
# Add tests and refactor as necessary

Add initial tests and refactor as necessary to make sure the tests pass.

Add the following tests:

- Tests for the `Symbols` utility class
- Tests for `Cells` utility class
- Add `CellGroups` test interfaces
- Add `Row` implementations test interface
- `Comparator` tests on `UniqueCellGroupTest`

   This is because we need to test the comparison characteristics of a
   `UniqueCellGroup` but since we are unable to extend the
   `ComparableTest` we add those tests directly to this interface.

- Tests for the `InterpolatableCellGroups` class.
- Tests for the `Block` class.
- Tests for the `Column` class.
- Tests for the `BoxBlock` class.
- Tests for the reference `Row` implementation.
- Tests for the reference `Column` implementation.
- Tests for the `InterpolatableCellGroups` class.
- Tests for the reference `BoxBlock` implementation.
- Tests for `CellGroups.rowOf`, `CellGroups.columnOf`, and `CellGroups.boxBlockOf` factory methods.

Included are also the following minor improvements and fixes:

- Add the `deepEquals` utility method in the `Cells` utility class.
- Rename the `swapCellPropeties` method to `swapCellSymbols` on the `Cells` class.
- Minor fix `Cells` utility class.
- Minor Javadoc fixes on the `UniqueCellGroup` test interface.
- Refactor `RowTest` by extending `InterpolatableCellGroupTest`.
- Refactor `BlockTest` to be generic.
- Limit the number of the final cells map to the size of `CellGroup`.
- Refactor reference `Row` and `Column` to accept valid cells only.
- Remove erroneous tests.
- Minor API fixes and Javadoc additions and clarifications.
- Add more clarifications of the name **Interpolatable**.
- Add a new invariant on `CellGroups.boxBlockOf`.

   The new invariant will ensure that the `BoxBlocks` instances returned by
   that factory method are valid by ensuring the product of the calculated
   `blockColumns` and `blockRows` are equal to the sizes of the new `BoxBlock`
   instances.
